cmd/uniq/uniq_test.go

```diff
@@ -0,0 +1,315 @@
+		},
+	},
+}
+
+var ttUniq = []struct {
+	opts     options
+	lines    []*Line
+	expected string
+}{
+	{
+		// uniq
+		options{},
+		ttScanLines[0].expected,
+		`world
+世
+3
+日本語
+`,
+	},
+	{
+		// uniq -num
+		options{printLineNumber: true},
+		ttScanLines[0].expected,
+		`2: world
+8: 世
+10: 3
+12: 日本語
+`,
+	},
+	{
+		// uniq -empty
+		options{printEmptyLines: true},
+		ttScanLines[1].expected,
+		`world
+
+世
+3
+日本語
+`,
+	},
+	{
+		// uniq -empty -num
+		options{
+			printEmptyLines: true,
+			printLineNumber: true,
+		},
+		ttScanLines[1].expected,
+		"2: world\n" +
+			"3,5: \n" +
+			`8: 世
+10: 3
+12: 日本語
+`,
+	},
+	{
+		// uniq -dup
+		options{printDuplicates: true},
+		ttScanLines[0].expected,
+		`hello
+世界
+1
+4
+`,
+	},
+	{
+		// uniq -dup -num
+		options{
+			printDuplicates: true,
+			printLineNumber: true,
+		},
+		ttScanLines[0].expected,
+		`1,4: hello
+6,7: 世界
+9,14: 1
+11,13: 4
+`,
+	},
+	{
+		// uniq -dup -empty
+		options{
+			printDuplicates: true,
+			printEmptyLines: true,
+		},
+		ttScanLines[1].expected,
+		`hello
+
+世界
+1
+4
+`,
+	},
+	{
+		// uniq -dup -empty -num
+		options{
+			printDuplicates: true,
+			printEmptyLines: true,
+			printLineNumber: true,
+		},
+		ttScanLines[1].expected,
+		"1,4: hello\n" +
+			"3,5: \n" +
+			`6,7: 世界
+9,14: 1
+11,13: 4
+`,
+	},
+
+	{
+		// uniq -every
+		options{printEveryOnce: true},
+		ttScanLines[0].expected,
+		`hello
+world
+世界
+世
+1
+3
+4
+日本語
+`,
+	},
+	{
+		// uniq -every -num
+		options{
+			printEveryOnce:  true,
+			printLineNumber: true,
+		},
+		ttScanLines[0].expected,
+		`1,4: hello
+2: world
+6,7: 世界
+8: 世
+9,14: 1
+10: 3
+11,13: 4
+12: 日本語
+`,
+	},
+	{
+		// uniq -every -empty
+		options{
+			printEveryOnce:  true,
+			printEmptyLines: true,
+		},
+		ttScanLines[1].expected,
+		`hello
+world
+
+世界
+世
+1
+3
+4
+日本語
+`,
+	},
+	{
+		// uniq -every -empty -num
+		options{
+			printEveryOnce:  true,
+			printEmptyLines: true,
+			printLineNumber: true,
+		},
+		ttScanLines[1].expected,
+		"1,4: hello\n" +
+			"2: world\n" +
+			"3,5: \n" +
+			`6,7: 世界
+8: 世
+9,14: 1
+10: 3
+11,13: 4
+12: 日本語
+`,
+	},
+}
+
+func cmpLines(a, b []*Line) error {
+	if la, lb := len(a), len(b); la != lb {
+		return fmt.Errorf("%T sizes diverge: %v != %v", a, la, lb)
+	}
+	for i, _ := range a {
+		linea := *a[i]
+		lineb := *b[i]
+		if *linea.text != *lineb.text {
+			return fmt.Errorf("line %d:\n%v !=\n%v",
+				i+1, *linea.text, *lineb.text)
+		}
+
+		if lena, lenb := len(linea.nums), len(lineb.nums); lena != lenb {
+			return fmt.Errorf("line counts diverge: %v != %v", lena, lenb)
+		}
+		for n, _ := range linea.nums {
+			lna := linea.nums[n]
+			lnb := lineb.nums[n]
+			if lna != lnb {
+				return fmt.Errorf("line number index: %d -"+
+					"numbers diverge: %v != %v", n, lna, lnb)
+			}
+		}
+	}
+	return nil
+}
+
+func TestScanLines(t *testing.T) {
+	for i, tScanLines := range ttScanLines {
+		bytesReader := bytes.NewReader(input)
+		scannedLines := scanLines(bytesReader, tScanLines.opts)
```

Comment:
As I'd said before, in case of error the entire test you exit.. Better to make this function return an error also.